Rewrite isValid() to explicit matches against a $valid$ field in the BMV2 backend

sethfowler opened the pull request:
We want to be able to use the P4Runtime serializer for any architecture and target. (This is issue #658.) There are currently two issues that need to be solved to do that:

1. We need to support extending the P4Runtime serializer with support for architecture-specific externs.
2. We need to remove all BMV2-specific code from the P4Runtime serializer.

This PR solves the second problem. At this point, the only BMV2-specific thing that's left is the implementation of `isValid()` in table keys. The P4Runtime serializer rewrites these to use either the `valid` match type, or to match against the special `$valid$` field that BMV2 adds to header types. This doesn't make sense to handle at that layer.

This PR adds a pass that adds the `$valid$` field to all header types and removes `isValid()` from the program. When `foo.isValid()` is used as a table key component, it's rewritten to `foo.$valid$`, which changes the type to `bit<1>`. In other contexts, it becomes `foo.$valid$ == 1`, which maintains the type of the overall expression. Because of the type change for key components, const table entries must also be rewritten to replace `true` and `false` with `1` and `0`.

This change lets us get rid of the `isValid()`-related code in the P4Runtime serializer. It also lets us remove similar code from the BMV2 JSON generator. Things become significantly simpler.

In addition to the tests in this repo, I tested this change with the `switch.p4` PTF tests. Everything works fine.

mbudiu-vmw commented on backends/bmv2/expression.cpp:
```diff
@@ -201,9 +188,9 @@ void ExpressionConverter::postorder(const IR::Member* expression)  {
-        if (field != nullptr)
-            // field could be a method call, i.e., isValid.
-            fieldName = field->controlPlaneName();
+        BUG_CHECK(field != nullptr,
```
There are other methods that look like fields, e.g., push.
So maybe this should not be a BUG.

sethfowler replied:
OK, I can change that.

mbudiu-vmw commented on backends/bmv2/synthesizeValidField.cpp:
```diff
@@ -0,0 +1,143 @@
+    }
+};
+
+/// Replaces calls to isValid() with references to the header's valid bit.
```
What about setValid?

sethfowler replied:
See the comments in the header. Currently we can't do that, because `setValid()` (when used in an action, at least) compiles down to the `add_header` primitive on BMV2. `add_header` does more than just set `$valid$` to 1 - for example, it's used in the implementation of header unions.

mbudiu-vmw replied:
isValid is also defined for unions.

sethfowler replied:
Ack, that's a potential problem! Let me take a look at that; I may need to handle that case differently.

sethfowler replied:
So I investigated this, and I don't think we're generating correct BMV2 JSON for header unions as it is. The `valid` match type won't work for header unions; we should be making use of the `valid_union` primitive. `valid_union` can't be used in table match keys (since they can't be general expressions) so what we need to do is use `valid_union` to set a scalar and use the scalar in the key. I think the right fix here is:

1. Don't remove `isValid()` calls on header unions in `SynthesizeValidField`. Since all other `isValid()` calls will be removed, these will be the only `isValid()` calls left.
2. Modify the `KeyIsComplex` policies to treat `isValid()` calls as complex, so they're moved out of match keys. (Or perhaps, for now, just add a new policy, and remove the old ones later once we're sure all backends have migrated over.)
3. In the backend, generate the `valid_union` primitive when we encounter `isValid()` calls on header unions.

Should all be pretty straightforward.

mbudiu-vmw replied:
Yes, we have a very broad issue covering the support of header unions in the back-end, which is very incomplete. I am working on this right now, but I don't know how much I will have time to finish before going on vacation. I will add the KeyIsComplex to the patch I working on now.

sethfowler replied:
Actually I already wrote the patch. =) I needed it as part of this PR. I'll add it in an additional commit, along with a test that exercises this functionality.

mbudiu-vmw commented on backends/bmv2/synthesizeValidField.cpp:
```diff
@@ -0,0 +1,143 @@
+    auto builtin = instance->to<P4::BuiltInMethod>();
+    if (builtin->name != IR::Type_Header::isValid) return expr;
+
+    // If isValid() is being used as a table key element, it already behaves
```
Why do you want to make it `bit<1>` instead of bool?
Anyway, if you insist it being a bit, then there is a cast from bit to bool which should be used instead of a comparison.

mbudiu-vmw replied:
Note that there are other perfectly legal expressions that evaluate to bool that can be used for table keys, so you still have to handle the other cases in the back-end.

sethfowler replied:
I actually did that because the P4-16 spec says that you can't have a boolean header field. However, now that I look, the BMV2 JSON generator already supports that, so perhaps it's not such a big deal. Making it a boolean field would eliminate the need for rewriting const table entries, which would be nice.

mbudiu-vmw replied:
The IR can represent programs that the spec disallows.
But indeed, the type-checker may be upset if it finds a boolean header field.

sethfowler replied:
Yeah, I just tried it and it doesn't pass the type checker. We could always loosen that check, though. I don't feel strongly either way, but I suppose that rewriting to something that's as close to valid P4 code as possible has some advantages.

mbudiu-vmw commented on backends/bmv2/synthesizeValidField.cpp:
```diff
@@ -0,0 +1,143 @@
+        if (key == nullptr) return table;
+        if (table->getEntries() == nullptr) return table;
+
+        // Find the indices of key elements which are simple references to valid
```
All this would be unnecessary if you don't change the type to bit.

sethfowler commented:
My thinking was that it didn't matter because field names aren't first class in P4Runtime.

It hasn't happened yet because we've had our heads down with other things, but I still expect us to add a P4-16 language feature to provide explicit names for match key components. With that in place, the idea is that you'd either provide an explicit or accept that the compiler will generate some arbitrary name for you. That's the same way we handle any other complex expression in a match key.

If there are problems with that approach, though, we can definitely revisit it.

antoninbas commented:
I think the name should be the same across all possible targets. For `isValid` it would be best IMO if it wasn't too arbitrary as I expect it to be pretty commonly used in match keys. So it would be nice if the name was predictable and not too shady. `<hdr>.$valid$` is not too bad. I would expect P4Runtime consumers to manipulate this name directly. A fully arbitrary name makes the annotation a requirement and I think we should avoid that and have a nice default name.

That's a secondary issue until we can find a way to truly make P4Runtime serializer target-independent by moving it up in the pass sequence.

sethfowler commented:
Sounds reasonable to me. We can standardize on `$valid$`, then, unless we find a reason to use something else.

For the general problem of making P4Runtime's output totally target independent, we're going to need to tackle the issue of control plane names more holistically in the P4-16 language itself. The good news is that we've come up with a lot of ideas for doing that; the bad news is that nobody's had time to finalize and implement them yet. =)

sethfowler commented:
@mbudiu-vmw could you please take a look one more time? Things are mostly the same, but I believe I've addressed your concern with how header unions are handled, and d40d956 implements the change we discussed that moves `isValid()` calls out of match keys in SimplifyKey.

I've also added a new integration test for matching against `isValid() on header unions (which would have caught the bug) and a reasonably thorough unit test suite for SynthesizeValidField.

mbudiu-vmw commented:
Note that I also submitted a PR which changes the key policy: #735

mbudiu-vmw commented on backends/bmv2/control.cpp:
```diff
@@ -342,32 +326,18 @@ ControlConverter::convertTable(const CFG::TableNode* node,
             keyelement->emplace("match_type", match_type);
-            auto jk = conv->convert(expr);
+
+            // XXX(seth): Boolean variables are stored as data, so we normally
```
Why the XXX?

sethfowler replied:
Because the name "convertLeftValue()" doesn't really capture the reason we're calling the function. It needs a better name, or perhaps we need a different function with a clearer name, but I couldn't immediately think of one.

mbudiu-vmw commented on backends/bmv2/synthesizeValidField.cpp:
```diff
@@ -0,0 +1,151 @@
+
+    // In other contexts, rewrite isValid() into a comparison with a constant.
+    // This maintains a boolean type for the overall expression.
+    auto constant = new IR::Constant(IR::Type::Bits::get(1), 1);
```
Why not Cast to bool - it would generate more efficient code in most cases it would be a noop, unlike the comparison.

sethfowler replied:
That's a good point. I'll make that change.

sethfowler replied:
Hmm, I just tried this and while the compiler is happy, it breaks several STF tests. The cast leads us to generate some code that malfunctions at runtime.

I've spent too long on this patch already so I'm going to go ahead and stick with the comparison for now, but I'll file a followup bug to investigate the issue.

mbudiu-vmw commented on backends/bmv2/synthesizeValidField.h:
```diff
@@ -0,0 +1,92 @@
+ * Calls to isValid() on header unions aren't rewritten; they can't be desugared
+ * to a simple read of a hidden field.
+ *
+ * XXX(seth): It would be cleaner to rewrite setValid() and setInvalid() too,
```
The XXX is probably not warranted.

sethfowler replied:
I actually do think we're eventually going to be able to do this. It requires some BMV2 changes and they aren't in place yet, but the documentation (search for `$valid$` [here](https://github.com/p4lang/behavioral-model/blob/master/docs/JSON_format.md)) suggests they will probably arrive eventually.
